donkeycar/pipeline/sequence.py

```diff
@@ -1,154 +1,90 @@
-                 x_transform: Callable[[R], XOut],
-                 y_transform: Callable[[R], YOut]) -> None:
-
+                 iterable: 'TfmShallowList') -> None:
         self.iterable = iterable
-        self.x_transform = x_transform
-        self.y_transform = y_transform
-        self.iterator = BaseTfmIterator_(
-            iterable=self.iterable, x_transform=self.x_transform, y_transform=self.y_transform)
+        self.iterator: Iterator[Union[TubRecord, Tuple[X, Y]]] \
+            = iter(self.iterable.iterable)
 
-    def __len__(self):
-        return len(self.iterator)
+    def __iter__(self) -> '_BaseTfmIterator[X, Y, XOut, YOut]':
+        return self
 
-    def __iter__(self) -> SizedIterator[Tuple[XOut, YOut]]:
-        return BaseTfmIterator_(
-            iterable=self.iterable, x_transform=self.x_transform, y_transform=self.y_transform)
+    def __next__(self) -> Tuple[XOut, YOut]:
+        r: Union[TubRecord, Tuple[X, Y]] = next(self.iterator)
+        return self.iterable.transform_item(r)
 
-    def __next__(self):
-        return next(self.iterator)
 
-
-class TfmTupleIterator(Generic[X, Y, XOut, YOut],  SizedIterator[Tuple[XOut, YOut]]):
+class TfmShallowList(Generic[XOut, YOut], SizedIterable[Tuple[XOut, YOut]]):
     def __init__(self,
-                 iterable: Iterable[Tuple[X, Y]],
-                 x_transform: Callable[[X], XOut],
-                 y_transform: Callable[[Y], YOut]) -> None:
-
+                 iterable: Union[List[TubRecord],
+                                 'TfmShallowList[XOut, YOut]'],
+                 x_transform: Callable[[Any], XOut],
+                 y_transform: Callable[[Any], YOut]) -> None:
         self.iterable = iterable
         self.x_transform = x_transform
         self.y_transform = y_transform
-        self.iterator = BaseTfmIterator_(
-            iterable=self.iterable, x_transform=self.x_transform, y_transform=self.y_transform)
-
-    def __len__(self):
-        return len(self.iterator)
 
-    def __iter__(self) -> SizedIterator[Tuple[XOut, YOut]]:
-        return BaseTfmIterator_(
-            iterable=self.iterable, x_transform=self.x_transform, y_transform=self.y_transform)
+    def __len__(self) -> int:
+        return len(self.iterable)
 
-    def __next__(self):
-        return next(self.iterator)
+    def __iter__(self) -> _BaseTfmIterator[X, Y, XOut, YOut]:
+        return _BaseTfmIterator(self)
 
+    def transform_item(self, r: Union[TubRecord, Tuple[X, Y]]) \
+            -> Tuple[XOut, YOut]:
+        if isinstance(r, tuple) and len(r) == 2:
+            x, y = r
+            return self.x_transform(x), self.y_transform(y)
+        else:
+            return self.x_transform(r), self.y_transform(r)
 
-class BaseTfmIterator_(Generic[XOut, YOut],  SizedIterator[Tuple[XOut, YOut]]):
-    '''
-    A basic transforming iterator.
-    Do no use this class directly.
-    '''
 
+class Pipeline(Generic[X, Y, XOut, YOut], SizedIterable[Tuple[XOut, YOut]]):
```

Comment:
This is a parallel implemention of the `TfmShallowList` with its `_BaseTfmIterator`. It provides the exact same functionality as the iterable / iterator protocol, by using the generator protocol.